File: dynamic-programming/494.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        return findsum(nums, 0, S);
    }
    
private:
    int findsum(vector<int>& nums, int p, int S) {
        if (nums.size() <= p) {
            if (S == 0) return 1;
            else return 0;
        }
        else {
            return findsum(nums, p+1, S-nums[p]) + findsum(nums, p+1, S+nums[p]);
        }
    }
};

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        int size = nums.size();
        result_ = vector<unordered_map<int, int>>(size, unordered_map<int, int>());
        return findsum(nums, 0, S);
    }
    
private:
    int findsum(vector<int>& nums, int p, int S) {
        if (nums.size() <= p) {
            if (S == 0) return 1;
            else return 0;
        }
        else {
            if (result_[p].count(S) != 0) {
                return result_[p][S];
            }
            else {
                result_[p][S] = findsum(nums, p+1, S-nums[p]) + findsum(nums, p+1, S+nums[p]);
                return result_[p][S];
            }
        }
    }
    
    vector<unordered_map<int, int>> result_;
};

File: dynamic-programming/494.java
class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        int sum = 0;
        for (int i = 0; i < nums.length; ++i) {
            sum += nums[i];
        }
        if ((sum + S) % 2 != 0) {
            return 0;
        }
        if (sum < Math.abs(S)) {
            return 0;
        }
        
        int target = (sum + S) / 2;
        // dp[i] is the number of combinations to get sum of i
        int[] dp = new int[target+1];
        dp[0] = 1;
        for (int num : nums) {
            for (int i = target; i >= num; --i) {
                dp[i] += dp[i-num];
            }
        }
        return dp[target];
    }
}
